File: app/core/session_store.py
"""
Chat session storage — JSON file-based implementation

Design:
- Repository Pattern: SessionRepository Protocol -> FileSessionStore
- One JSON file per session: output/sessions/{session_id}.json
- Append-only messages: read + append + write
- Future: swap to Supabase with zero router change
"""
from __future__ import annotations

import html
import json
import logging
import re
import threading
import uuid
from collections import defaultdict
from dataclasses import asdict, dataclass, field
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional, Protocol, Tuple

from app import config

logger = logging.getLogger(__name__)

MAX_MESSAGES_PER_SESSION = 100

_UUID_RE = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$"
)

_session_locks: dict = defaultdict(threading.Lock)


def _now_iso() -> str:
    now = datetime.now(timezone.utc)
    return now.strftime("%Y-%m-%dT%H:%M:%S.") + f"{now.microsecond // 1000:03d}Z"


def _sanitize_title(raw: str) -> str:
    """Sanitize user content for use as a session title."""
    return html.escape(raw[:50]).strip()


@dataclass(frozen=True)
class SessionMessage:
    role: str  # "user" | "assistant"
    content: str
    sources: list[dict] = field(default_factory=list)
    created_at: str = field(default_factory=_now_iso)


@dataclass
class Session:
    id: str
    title: str
    messages: list[SessionMessage] = field(default_factory=list)
    created_at: str = field(default_factory=_now_iso)
    updated_at: str = field(default_factory=_now_iso)

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "title": self.title,
            "messages": [asdict(m) for m in self.messages],
            "created_at": self.created_at,
            "updated_at": self.updated_at,
        }

    @classmethod
    def from_dict(cls, data: dict) -> Session:
        messages = [SessionMessage(**m) for m in data.get("messages", [])]
        return cls(
            id=data["id"],
            title=data.get("title", ""),
            messages=messages,
            created_at=data.get("created_at", ""),
            updated_at=data.get("updated_at", ""),
        )


class SessionRepository(Protocol):
    def list_sessions(
        self, limit: int = 20, offset: int = 0
    ) -> Tuple[list[Session], int]: ...

    def get_session(self, session_id: str) -> Optional[Session]: ...

    def create_session(self, title: str = "") -> Session: ...

    def add_message(
        self, session_id: str, msg: SessionMessage
    ) -> Optional[Session]: ...

    def delete_session(self, session_id: str) -> bool: ...


class FileSessionStore:
    """JSON file-based session storage: output/sessions/{session_id}.json"""

    def __init__(self, base_dir: Optional[Path] = None) -> None:
        self._base_dir = base_dir or config.SESSIONS_DIR
        self._base_dir.mkdir(parents=True, exist_ok=True)

    def _validate_id(self, session_id: str) -> None:
        """Validate session_id is a strict UUID v4 to prevent path traversal."""
        if not _UUID_RE.match(session_id):
            raise ValueError(f"Invalid session_id format: {session_id!r}")

    def _path(self, session_id: str) -> Path:
        self._validate_id(session_id)
        return self._base_dir / f"{session_id}.json"

    def _read(self, session_id: str) -> Optional[Session]:
        try:
            path = self._path(session_id)
        except ValueError:
            return None
        if not path.exists():
            return None
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
            return Session.from_dict(data)
        except (json.JSONDecodeError, KeyError, TypeError) as exc:
            logger.warning("session_store: read failed %s — %s", path, exc)
            return None

    def _write(self, session: Session) -> None:
        path = self._path(session.id)
        try:
            path.parent.mkdir(parents=True, exist_ok=True)
            path.write_text(
                json.dumps(session.to_dict(), ensure_ascii=False, indent=2),
                encoding="utf-8",
            )
        except OSError as exc:
            logger.warning("session_store: write failed %s — %s", path, exc)

    def list_sessions(
        self, limit: int = 20, offset: int = 0
    ) -> Tuple[list[Session], int]:
        files = sorted(
            self._base_dir.glob("*.json"),
            key=lambda f: f.stat().st_mtime,
            reverse=True,
        )
        # Parse all files, skip corrupt ones, count only valid sessions
        all_sessions = [
            s for f in files if (s := self._read(f.stem)) is not None
        ]
        total = len(all_sessions)
        return all_sessions[offset : offset + limit], total

    def get_session(self, session_id: str) -> Optional[Session]:
        return self._read(session_id)

    def create_session(self, title: str = "") -> Session:
        session = Session(
            id=str(uuid.uuid4()),
            title=_sanitize_title(title) if title else "New Chat",
        )
        self._write(session)
        return session

    def add_message(
        self, session_id: str, msg: SessionMessage
    ) -> Optional[Session]:
        with _session_locks[session_id]:
            session = self._read(session_id)
            if session is None:
                return None
            if len(session.messages) >= MAX_MESSAGES_PER_SESSION:
                logger.warning(
                    "session_store: session %s reached max messages (%d)",
                    session_id,
                    MAX_MESSAGES_PER_SESSION,
                )
                return None
            # Immutable update: create new list
            new_title = session.title
            if new_title == "New Chat" and msg.role == "user":
                new_title = _sanitize_title(msg.content)

            updated = Session(
                id=session.id,
                title=new_title,
                messages=[*session.messages, msg],
                created_at=session.created_at,
                updated_at=_now_iso(),
            )
            self._write(updated)
            return updated

    def delete_session(self, session_id: str) -> bool:
        try:
            path = self._path(session_id)
        except ValueError:
            return False
        if not path.exists():
            return False
        try:
            path.unlink()
            return True
        except OSError as exc:
            logger.warning("session_store: delete failed %s — %s", path, exc)
            return False


# Module-level singleton
session_store = FileSessionStore()

File: api/src/store/session-store.ts
/**
 * File-based session store — one JSON file per session.
 *
 * Translated from Python app/core/session_store.py
 */

import {
  readFileSync,
  writeFileSync,
  readdirSync,
  unlinkSync,
  existsSync,
  mkdirSync,
  statSync,
} from "node:fs";
import { join } from "node:path";
import { v4 as uuidv4 } from "uuid";
import { paths } from "../config.js";
import { escapeHtml } from "../utils/sanitize.js";

const MAX_MESSAGES_PER_SESSION = 100;

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;

function nowIso(): string {
  const now = new Date();
  return now.toISOString().replace(/(\.\d{3})\d*Z$/, "$1Z");
}

function sanitizeTitle(raw: string): string {
  return escapeHtml(raw.slice(0, 50)).trim();
}

export interface SessionMessage {
  readonly role: string;
  readonly content: string;
  readonly sources: readonly Record<string, unknown>[];
  readonly created_at: string;
  readonly metadata?: Record<string, unknown>;
}

export interface Session {
  readonly id: string;
  readonly title: string;
  readonly messages: readonly SessionMessage[];
  readonly created_at: string;
  readonly updated_at: string;
}

export class FileSessionStore {
  private readonly baseDir: string;

  constructor(baseDir?: string) {
    this.baseDir = baseDir ?? paths.sessionsDir;
    if (!existsSync(this.baseDir)) {
      mkdirSync(this.baseDir, { recursive: true });
    }
  }

  private validateId(sessionId: string): void {
    if (!UUID_RE.test(sessionId)) {
      throw new Error(`Invalid session_id format: ${sessionId}`);
    }
  }

  private filePath(sessionId: string): string {
    this.validateId(sessionId);
    return join(this.baseDir, `${sessionId}.json`);
  }

  private readSession(sessionId: string): Session | null {
    try {
      const fp = this.filePath(sessionId);
      if (!existsSync(fp)) return null;
      const data = JSON.parse(readFileSync(fp, "utf-8"));
      return data as Session;
    } catch {
      return null;
    }
  }

  private writeSession(session: Session): void {
    const fp = this.filePath(session.id);
    writeFileSync(fp, JSON.stringify(session, null, 2), "utf-8");
  }

  listSessions(
    limit: number = 20,
    offset: number = 0,
  ): { sessions: readonly Session[]; total: number } {
    if (!existsSync(this.baseDir)) return { sessions: [], total: 0 };

    const files = readdirSync(this.baseDir)
      .filter((f) => f.endsWith(".json"))
      .map((f) => ({
        name: f,
        mtime: statSync(join(this.baseDir, f)).mtimeMs,
      }))
      .sort((a, b) => b.mtime - a.mtime);

    const allSessions: Session[] = [];
    for (const { name } of files) {
      const stem = name.replace(".json", "");
      const s = this.readSession(stem);
      if (s) allSessions.push(s);
    }

    return {
      sessions: allSessions.slice(offset, offset + limit),
      total: allSessions.length,
    };
  }

  getSession(sessionId: string): Session | null {
    return this.readSession(sessionId);
  }

  createSession(title: string = ""): Session {
    const session: Session = {
      id: uuidv4(),
      title: title ? sanitizeTitle(title) : "New Chat",
      messages: [],
      created_at: nowIso(),
      updated_at: nowIso(),
    };
    this.writeSession(session);
    return session;
  }

  addMessage(sessionId: string, msg: SessionMessage): Session | null {
    const session = this.readSession(sessionId);
    if (!session) return null;

    if (session.messages.length >= MAX_MESSAGES_PER_SESSION) {
      console.warn(
        `session_store: session ${sessionId} reached max messages (${MAX_MESSAGES_PER_SESSION})`,
      );
      return null;
    }

    let newTitle = session.title;
    if (newTitle === "New Chat" && msg.role === "user") {
      newTitle = sanitizeTitle(msg.content);
    }

    const updated: Session = {
      id: session.id,
      title: newTitle,
      messages: [...session.messages, msg],
      created_at: session.created_at,
      updated_at: nowIso(),
    };
    this.writeSession(updated);
    return updated;
  }

  deleteSession(sessionId: string): boolean {
    try {
      const fp = this.filePath(sessionId);
      if (!existsSync(fp)) return false;
      unlinkSync(fp);
      return true;
    } catch {
      return false;
    }
  }
}

import { hasSupabase } from "./supabase-client.js";
import { SupabaseSessionStore } from "./supabase-session-store.js";

/**
 * Unified async session store interface.
 * FileSessionStore methods are sync but wrapped in Promise for compatibility.
 */
export interface AsyncSessionStore {
  listSessions(limit?: number, offset?: number): Promise<{ sessions: readonly Session[]; total: number }>;
  getSession(sessionId: string): Promise<Session | null>;
  createSession(title?: string): Promise<Session>;
  addMessage(sessionId: string, msg: SessionMessage): Promise<Session | null>;
  deleteSession(sessionId: string): Promise<boolean>;
}

function wrapFileStore(fs: FileSessionStore): AsyncSessionStore {
  return {
    listSessions: (limit, offset) => Promise.resolve(fs.listSessions(limit, offset)),
    getSession: (id) => Promise.resolve(fs.getSession(id)),
    createSession: (title) => Promise.resolve(fs.createSession(title)),
    addMessage: (id, msg) => Promise.resolve(fs.addMessage(id, msg)),
    deleteSession: (id) => Promise.resolve(fs.deleteSession(id)),
  };
}

/** Factory: returns SupabaseSessionStore or FileSessionStore (async wrapper). */
export function createSessionStore(): AsyncSessionStore {
  if (hasSupabase()) {
    console.log("SessionStore: using Supabase");
    return new SupabaseSessionStore();
  }
  console.log("SessionStore: using file-based");
  return wrapFileStore(new FileSessionStore());
}

export const sessionStore = createSessionStore();
